packages/devtools_app/lib/src/eval_on_dart_library.dart

```
@@ -156,25 +178,229 @@ class EvalOnDartLibrary {
+      '    final result = $expression;'
+      '    reader.add(result);'
+      '  } catch (err, stack) {'
+      '    reader.add(err);'
+      '    reader.add(stack);'
+      '  } finally {'
+      '    postEvent("future_completed", {"id": $id});'
+      '  }'
+      '}()',
+      isAlive: isAlive,
+      scope: {
+        ...?scope,
+        'postEvent': postEventRef.id,
+        'widgetInspectorService': widgetInspectorServiceRef.id,
+      },
+    );
+
+    await future;
+
+    final resultRef = await evalInstance(
+      '() {'
+      '  final result = widgetInspectorService.toObject("$readerId", "$readerGroup") as List;'
+      '  widgetInspectorService.disposeGroup("$readerGroup");'
+      '  return result;'
+      '}()',
+      isAlive: isAlive,
+      scope: {'widgetInspectorService': widgetInspectorServiceRef.id},
+    );
+
+    assert(resultRef.length == 1 || resultRef.length == 2);
+    if (resultRef.length == 2) {
+      throw FutureFailedException(
+        expression,
+        resultRef.elements[0],
+        resultRef.elements[1],
+      );
+    }
+
+    return resultRef.elements[0];
+  }
+
+  /// An [eval] that throws when a [Sentinel]/error occurs or if [isAlive] was
+  /// disposed while the request was pending.
+  ///
+  /// If `isAlive` was disposed while the request was pending, will throw a [CancelledException].
+  Future<InstanceRef> safeEval(
+    String expression, {
+    @required Disposable isAlive,
+    Map<String, String> scope,
+  }) async {
+    Object result;
+
+    try {
+      if (disposed) {
+        throw StateError(
+          'Called `safeEval` on a disposed `EvalOnDartLibrary` instance',
+        );
+      }
+
+      result = await addRequest(isAlive, () async {
+        final libraryRef = await _waitForLibraryRef();
+
+        return await service.evaluate(
+          isolateId,
+          libraryRef.id,
+          expression,
+          scope: scope,
+        );
+      });
+
+      if (result == null) return null;
```

this case is where the EvalOnDartLibrary was disposed while the request was in progress. Do you want to throw an error here or only if it was disposed before the request started?